font/src/directwrite/mod.rs

```diff
@@ -0,0 +1,222 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//! Rasterization powered by DirectWrite
+extern crate dwrote;
+use self::dwrote::{
+    FontCollection, FontStretch, FontStyle, FontWeight, GlyphOffset, GlyphRunAnalysis,
+};
+
+use super::{FontDesc, FontKey, GlyphKey, Metrics, RasterizedGlyph, Size, Slant, Style, Weight};
+
+pub struct DirectWriteRasterizer {
+    fonts: Vec<dwrote::FontFace>,
+    device_pixel_ratio: f32,
+}
+
+impl crate::Rasterize for DirectWriteRasterizer {
+    type Err = Error;
+
+    fn new(device_pixel_ratio: f32, _: bool) -> Result<DirectWriteRasterizer, Error> {
+        Ok(DirectWriteRasterizer { fonts: Vec::new(), device_pixel_ratio })
+    }
+
+    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
+        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+
+        let vmetrics = font.metrics();
+        let scale = (size.as_f32_pts() * self.device_pixel_ratio * (96.0 / 72.0))
+            / f32::from(vmetrics.designUnitsPerEm);
+
+        let underline_position = f32::from(vmetrics.underlinePosition) * scale;
+        let underline_thickness = f32::from(vmetrics.underlineThickness) * scale;
+
+        let strikeout_position = f32::from(vmetrics.strikethroughPosition) * scale;
+        let strikeout_thickness = f32::from(vmetrics.strikethroughThickness) * scale;
+
+        let line_height = f64::from(
+            i32::from(vmetrics.ascent) + i32::from(vmetrics.descent) + i32::from(vmetrics.lineGap),
+        ) * f64::from(scale);
+        let descent = -f32::from(vmetrics.descent) * scale;
+
+        // Similarly to rusttype we assume that all monospace characters have the same width
+        // Because of this we take 33, '!', the first drawable character for measurements
+        let glyph_metrics = font.get_design_glyph_metrics(&[33], false);
+        let hmetrics = glyph_metrics.first().ok_or(Error::MissingGlyph('!'))?;
+
+        let average_advance = f64::from(hmetrics.advanceWidth) * f64::from(scale);
+
+        Ok(Metrics {
+            descent,
+            average_advance,
+            line_height,
+            underline_position,
+            underline_thickness,
+            strikeout_position,
+            strikeout_thickness,
+        })
+    }
+
+    fn load_font(&mut self, desc: &FontDesc, _size: Size) -> Result<FontKey, Error> {
+        let system_fc = FontCollection::system();
+
+        let weight = match desc.style {
+            Style::Description { weight, .. } => match weight {
+                Weight::Normal => FontWeight::Regular,
+                Weight::Bold => FontWeight::Bold,
+            },
+
+            _ => FontWeight::Regular,
+        };
```

Comment:
Rust's pattern matching actually has quite a neat trick to simplify this even more, I think this should work:

```rust
let style = if let Style::Description { weight: Weight::Bold, .. } = desc.style {
    FontWeight::Bold
} else {
    FontWeight::Regular
};
```